File: 3. First Negative Number in Every Window of Size K.py
"""
First Negative Number in Every Window of Size K
"""
from collections import deque

def first_negative(arr, k):
    dq = deque()
    n = len(arr)
    res = []
    for i in range(k):
        if arr[i] < 0: dq.append(i)
    for i in range(k, n):
        res.append(arr[dq[0]] if dq else 0)
        while dq and dq[0] <= i-k: dq.popleft()
        if arr[i] < 0: dq.append(i)
    res.append(arr[dq[0]] if dq else 0)
    return res

def main():
    arr = [12,-1,-7,8,-15,30,16,28]
    print(first_negative(arr,3))

if __name__ == '__main__':
    main()

# Time Complexity: O(n)
# Space Complexity: O(k)

File: 3. First Negative Number in Every Window of Size K.cpp
// Topic: First Negative Number in Every Window of Size K
#include <bits/stdc++.h>
using namespace std;

int main()
{
    vector<int> arr = {12, -1, -7, 8, -15, 30, 16, 28};
    int k = 3;
    deque<int> dq;  // store indices of negative numbers
    int n = arr.size();
    for (int i = 0; i < k; i++)
        if (arr[i] < 0)
            dq.push_back(i);
    for (int i = k; i < n; i++)
    {
        if (!dq.empty())
            cout << arr[dq.front()] << ' ';
        else
            cout << 0 << ' ';
        while (!dq.empty() && dq.front() <= i - k) dq.pop_front();
        if (arr[i] < 0)
            dq.push_back(i);
    }
    if (!dq.empty())
        cout << arr[dq.front()];
    else
        cout << 0;
    cout << '\n';
    return 0;
}

// Time Complexity: O(n)
// Space Complexity: O(k)
